### PR TITLE
feat(runtime): white list some always in memory contract

### PR DESCRIPTION
**Pull request by ailisp**
For example, near_evm contract. It's determined by contract's account id.

Test Plan
---------
near vm runner tests; 
doing debug print as in `cache-slow-repro-whitelist` branch and see no deserialization caused latency even when in memory cache is disabled for near_evm.wasm contract. And when it's not in whitelist, this deserialization latency happens, as before.

### CODE REVIEW

**Comment by nearmax** [comment]
I suggest to extract `ALWAYS_IN_MEM_CONTRACT` into a config file.

File: runtime/near-vm-runner/src/cache.rs [hunk 1]
```diff
@@ -3,16 +3,55 @@ use crate::prepare;
...
 use wasmer_runtime_core::load_cache_with;
 
+lazy_static! {
+    static ref ALWAYS_IN_MEM_CONTRACT: RwLock<HashMap<AccountId, Option<(CryptoHash, Result<wasmer_runtime::Module, VMError>)>>> = {
+        let mut m = HashMap::new();
+        m.insert("evm".to_string(), None);
+        // Add more always in memory contract account ids
+        RwLock::new(m)
+    };
+}
```
Comment (matklad) [inline runtime/near-vm-runner/src/cache.rs:27]: Software-architecture wise, I am somewhat surprised that we use a global variable here, instead of storing this as a field of some kind of context struct. As we might want to have several runtime instances (for sharding), I fear that this'll bite us down the road. But this isn't specific to the PR, the `MODULES` below suffer from the same problem.

Comment (olonho) [reply runtime/near-vm-runner/src/cache.rs:27]: Overall it doesn't look like proper long-term approach. 
First: the name is not proper contract identifier, content hash is.
Second: this kind of information belongs to configuration, not to the binary code.

Comment (ailisp) [reply runtime/near-vm-runner/src/cache.rs:27]: Yes it's problematic for multiple runner instance, existing code assumed there's only one runner

**Comment by ailisp** [comment]
> I suggest to extract ALWAYS_IN_MEM_CONTRACT into a config file.

@nearmax I suppose this goes to config.json instead of genesis config, to not introduce protocol change

File: runtime/near-vm-runner/src/cache.rs (hunk 1 above)
Comment (bowenwang1996) [reply runtime/near-vm-runner/src/cache.rs:27]: As @matklad mentioned, it becomes a problem when we introduce parallel runtimes, which may be very soon

File: neard/src/runtime.rs [hunk 2]
```diff
@@ -128,6 +129,7 @@ pub struct NightshadeRuntime {
     genesis_runtime_config: Arc<RuntimeConfig>,
 
     store: Arc<Store>,
+    in_memory_contracts: InMemoryContracts,
```
Comment (bowenwang1996) [inline neard/src/runtime.rs:132]: why don't we make it per shard to reduce potentially lock contention?

File: neard/src/config.rs [hunk 3]
```diff
@@ -550,6 +552,7 @@ pub struct NearConfig {
     pub client_config: ClientConfig,
     pub network_config: NetworkConfig,
     pub rpc_config: RpcConfig,
+    pub cache_config: CacheConfig,
```
Comment (matklad) [inline neard/src/config.rs:555]: This is a too general name for this context. Ie, it's not clear it cache applies to chain/network/vm. `VMCacheConfig` would be better.

As an aside, it actually should be `VmCacheConfig`, according to Rust naming rules, but we have `VM` everywhere, so I guess its better to roll with that.

File: runtime/near-vm-runner/src/runner.rs [hunk 4]
```diff
@@ -30,6 +31,7 @@ pub fn run<'a>(
     promise_results: &'a [PromiseResult],
     current_protocol_version: ProtocolVersion,
     cache: Option<&'a dyn CompiledContractCache>,
+    in_mem_contract: InMemoryContracts,
```
Comment (matklad) [inline runtime/near-vm-runner/src/runner.rs:34]: General observation: function with many arguments are also functions with frequently changing set of arguments. We should look into packing some of the args here into a struct, to reduce code churn:

https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#functions-with-many-parameters

but this is unrelated to the PR

File: neard/src/runtime.rs (hunk 2 above)
Comment (matklad) [reply neard/src/runtime.rs:132]: I'd probably suggest to YAGNI that, unless the code is proven guilty for a perf regression. With per-shard, we'll be using N times more memory storing the same bytes. I think a more promising optimization avenue here (should we need it) is to precompile stuff at stratup, and then share a read-only cache.

File: core/primitives/src/runtime/in_memory_contract.rs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::sync::Arc;
+use crate::{
+    hash::CryptoHash,
+    types::AccountId,
+};
+use near_vm_errors::VMError;
+
+pub type InMemoryContracts = Arc<RwLock<HashMap<AccountId, Option<(CryptoHash, Result<wasmer_runtime::Module, VMError>)>>>>;
```
Comment (matklad) [inline core/primitives/src/runtime/in_memory_contract.rs:10]: Hm, taking a bird's eye view, the fact fact the datastructure for in-memory cache which is used by the vm-runner crate, is defined outside the vm-runner crate feels wrong. 

I guess it's reasonable place to do this, given the current structure of code. But the structure itself is suspicious .

Comment (ailisp) [reply core/primitives/src/runtime/in_memory_contract.rs:10]: this should be struct InMemoryContracts(Arc ...) instead of type. but it should be here because primitivies/src/runtime/apply_state.rs is the place to use it, if it were in runtime crate there would be an additional circular dependency primitives depends on runtime

File: neard/src/config.rs (hunk 3 above)
Comment (ailisp) [reply neard/src/config.rs:555]: Sounds good

File: neard/src/runtime.rs (hunk 2 above)
Comment (ailisp) [reply neard/src/runtime.rs:132]: Yeah, make it per shard sounds good

File: runtime/near-vm-runner/src/runner.rs (hunk 4 above)
Comment (ailisp) [reply runtime/near-vm-runner/src/runner.rs:34]: Will do in later refactor PRs, you're also welcome to do, your refactor PRs look very good

File: neard/src/config.rs (hunk 3 above)
Comment (ailisp) [reply neard/src/config.rs]: renamed

File: neard/src/runtime.rs (hunk 2 above)
Comment (ailisp) [reply neard/src/runtime.rs]: Github didn't show @matklad 's comment when i comment previously. I thinks per shard mem is better. We don't try to tell whether it's a same bytes before cache it (it's cached based on account name, not contract bytes), so won't spend more memory.

File: runtime/near-vm-runner/src/cache.rs (hunk 1 above)
Comment (ailisp) [reply runtime/near-vm-runner/src/cache.rs]: Has been redo to keep in mem in NightshadeRuntime, and pass in as arguments

File: core/primitives/src/runtime/in_memory_contract.rs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        match self.0.read().unwrap().get(account_id) {
+            Some(Some((k, r))) => {
+                if k == key {
+                    Some(Some(r.clone()))
+                } else {
+                    None
+                }
+            }
+            Some(None) => Some(None),
+            None => None,
```
Comment (EgorKulikov) [inline core/primitives/src/runtime/in_memory_contract.rs:43]: Why do we return None when there is a contract with another key and when there is no such account_id, but Some(None) when there is no contract for this account? This seem weird

File: neard/src/runtime.rs (hunk 2 above)
Comment (bowenwang1996) [reply neard/src/runtime.rs]: @matklad yeah there will be no repetition yet unless we introduce contracts that are globally accessible/

File: core/primitives/src/runtime/in_memory_contract.rs (hunk 6 above)
Comment (EgorKulikov) [reply core/primitives/src/runtime/in_memory_contract.rs:43]: it may well be intended, but I think it is at least worth documenting

File: core/primitives/src/runtime/in_memory_contract.rs [hunk 7]
```diff
@@ -0,0 +1,56 @@
+use crate::{hash::CryptoHash, types::AccountId};
+use near_vm_errors::VMError;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::sync::RwLock;
+
+#[derive(Clone)]
+pub struct InMemoryContracts(
+    pub  Arc<
+        RwLock<HashMap<AccountId, Option<(CryptoHash, Result<wasmer_runtime::Module, VMError>)>>>,
```
Comment (bowenwang1996) [inline core/primitives/src/runtime/in_memory_contract.rs:10]: it is better to extract the value of the hashmap into a separate type.

**Review by bowenwang1996 — COMMENTED** [review]
I don't think we should use the same cache for both client and view client. We should either use no cache for view client or use a separate cache.

File: core/primitives/src/runtime/in_memory_contract.rs (hunk 6 above)
Comment (ailisp) [reply core/primitives/src/runtime/in_memory_contract.rs:43]: Good catch, it should be Some(None) too. I also add a doc string

**Comment by ailisp** [comment]
@bowenwang1996 
this pr didn't change view_client, do you mean state_viewer? if so state_viewer is an ops tool, I think better to pass an empty contract cache, to avoid making main api `run_vm` etc takes Option<InMemoryContracts> (add additional `Option`). Use a separate cache for state_viewer turns this into a trait, also add unnecessary cost to `run_vm`.

**Review by EgorKulikov — APPROVED** [review]

File: core/primitives/src/runtime/in_memory_contract.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            map.insert(account_id.clone(), None);
+        }
+        Self(Arc::new(RwLock::new(map)))
+    }
+
+    /// Get contract in InMemoryContracts cache
+    /// Returns Some(ContractResult) when account is in whitelist and contract cache is valid
+    /// Returns Some(None) when account is in whitelist but contract cache isn't exist (first compile),
+    /// or invalid (another contract being deployed and compiled)
+    /// Returns None when account is not in whitelist
```
Comment (frol) [inline core/primitives/src/runtime/in_memory_contract.rs:33]: Can we actually just use a custom enum with 3 variants?

**Comment by bowenwang1996** [comment]
> this pr didn't change view_client, do you mean state_viewer?

Oh I see you already pass the empty cache here https://github.com/near/nearcore/pull/3972/files#diff-9811f5bb9cc495ab24e1f7ab510a6ff683aeef10dee2a59698f3dd05fdbb08c5R146

File: neard/src/runtime.rs (hunk 2 above)
Comment (matklad) [reply neard/src/runtime.rs]: Ack! might make sense to shard this from the start then, if this doesn't make the code more complex.

File: runtime/near-vm-runner/Cargo.toml [hunk 9]
```diff
@@ -26,6 +26,7 @@ near-primitives = { path = "../../core/primitives" }
 log = "0.4"
 near-evm-runner = { path = "../near-evm-runner", optional = true }
 cached = "0.23.0"
+lazy_static = "1.4"
```
Comment (matklad) [inline runtime/near-vm-runner/Cargo.toml:29]: No longer needed?

File: core/primitives/src/runtime/in_memory_contract.rs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+pub struct InMemoryContracts(pub Arc<RwLock<HashMap<AccountId, InMemoryContractEntry>>>);
+
+impl std::fmt::Debug for InMemoryContracts {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "InMemoryContracts")
+    }
+}
+
+impl InMemoryContracts {
+    pub fn new(in_memory_account_ids: &Vec<AccountId>) -> Self {
```
Comment (matklad) [inline core/primitives/src/runtime/in_memory_contract.rs:20]:
```suggestion
    pub fn new(in_memory_account_ids: &[AccountId]) -> Self {
```

`&Vec<T>` is useless type: https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#useless-types. 

`s/&vec![]/&[]/g` after this

**Comment by bowenwang1996** [comment]
@ailisp do we still want to proceed with this PR?

**Comment by ailisp** [comment]
@bowenwang1996 I think let's keep it open. If 1) there's prove that wasmer2 deserialize fast enough that we can drop in memory cache, and 2) we've fully tested wasmer2 runner to supersede wasmer 0.17, then we're sure we don't need this PR.

If there is N+1 number of active contracts on mainnet that evict evm contract frequently, before we're ready with 1) 2) then we will need this.

**Comment by bowenwang1996** [comment]
> If there is N+1 number of active contracts on mainnet that evict evm contract frequently, before we're ready with 1) 2) then we will need this.

How exactly does this help? The fees are the same

**Comment by ailisp** [comment]
@bowenwang1996 the problem is not fee, it's when evm contract not in memory cache deserialization takes too long

**Comment by ailisp** [comment]
ping
